// main-menu/build/menu/MenuItem.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _reactEs6Component = require('react-es6-component');

var _reactEs6Component2 = _interopRequireDefault(_reactEs6Component);

var _mousetrap = require('mousetrap');

var _mousetrap2 = _interopRequireDefault(_mousetrap);

var MenuItem = (function (_Component) {
  _inherits(MenuItem, _Component);

  _createClass(MenuItem, null, [{
    key: 'propTypes',
    value: {
      label: _react2['default'].PropTypes.string,
      separator: _react2['default'].PropTypes.bool,
      shortcut: _react2['default'].PropTypes.string,
      onClose: _react2['default'].PropTypes.func,
      onClick: _react2['default'].PropTypes.func
    },
    enumerable: true
  }, {
    key: 'defaultProps',
    value: {
      onClose: function onClose() {},
      onClick: function onClick() {}
    },
    enumerable: true
  }]);

  function MenuItem() {
    var _this = this;

    _classCallCheck(this, MenuItem);

    _get(Object.getPrototypeOf(MenuItem.prototype), 'constructor', this).apply(this, arguments);
    this.state = {
      open: false
    };
    this._isOpen = false;
    if (this.props.shortcut) {
      _mousetrap2['default'].bind(this.props.shortcut.toLowerCase(), function () {
        _this.props.onClick();
      });
    }
  }

  _createClass(MenuItem, [{
    key: 'render',
    value: function render() {
      if (this.props.separator) {
        return _react2['default'].createElement(
          'div',
          { className: 'menu-item-separator' },
          _react2['default'].createElement('div', { className: 'sep' })
        );
      } else {
        var className = "menu-item";
        if (this.state.open) {
          className += " open";
        }
        return _react2['default'].createElement(
          'div',
          _extends({ onMouseEnter: this._onMouseEnter,
            onMouseLeave: this._onMouseLeave,
            onClickCapture: this._onMenuClick,
            className: className }, this.props),
          _react2['default'].createElement(
            'div',
            { className: 'icon' },
            this._renderIcon()
          ),
          _react2['default'].createElement(
            'span',
            { className: 'menu-title' },
            this.props.label
          ),
          _react2['default'].createElement(
            'span',
            { className: 'menu-shortcut' },
            this.props.shortcut
          ),
          this._renderExpanderArrow(),
          this._renderChildren()
        );
      }
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this._addListeners();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this._removeListeners();
    }
  }, {
    key: 'isOpen',
    value: function isOpen() {
      return this._isOpen;
    }
  }, {
    key: 'open',
    value: function open() {
      this._isOpen = true;
      this.setState({
        open: true
      });
    }
  }, {
    key: 'close',
    value: function close() {
      this._isOpen = false;
      this.setState({
        open: false
      });
    }
  }, {
    key: '_renderIcon',
    value: function _renderIcon() {
      if (this.props.icon) {
        return _react2['default'].createElement('img', { src: this.props.icon });
      }
    }
  }, {
    key: '_renderExpanderArrow',
    value: function _renderExpanderArrow() {
      if (this.props.children) {
        return _react2['default'].createElement('div', { className: 'arrow-right' });
      }
    }
  }, {
    key: '_renderChildren',
    value: function _renderChildren() {
      var _this2 = this;

      if (this.props.children) {
        var children = _react2['default'].Children.map(this.props.children, function (element) {
          return _react2['default'].cloneElement(element, { onClose: _this2._onCloseChild });
        });
        return _react2['default'].createElement(
          'div',
          { className: 'menu-children' },
          children
        );
      }
    }
  }, {
    key: '_onMouseEnter',
    value: function _onMouseEnter() {
      var _this3 = this;

      clearTimeout(this._menuCloseTimeout);
      this._menuItemTimeout = setTimeout(function () {
        _this3._hoverClick = true;
        _this3._isOpen = true;
        _this3.setState({
          open: true
        });
      }, 500);
    }
  }, {
    key: '_onMouseLeave',
    value: function _onMouseLeave(event) {
      var _this4 = this;

      clearTimeout(this._menuItemTimeout);
      this._menuCloseTimeout = setTimeout(function () {
        if (_this4._hoverClick) {
          _this4._hoverClick = false;
          _this4._isOpen = false;
          _this4.setState({
            open: false
          });
        }
      }, 200);
    }
  }, {
    key: '_onMenuClick',
    value: function _onMenuClick(event) {

      this._hoverClick = false;
      clearTimeout(this._menuItemTimeout);
      if (this._isLeaf()) {
        event.preventDefault();
        event.stopPropagation();
        this._isOpen = false;
        this.setState({
          open: false
        });
        this.props.onClick();
        if (this.props.onClose) {
          this.props.onClose();
        }
      } else {
        if (!this.state.open) {
          event.preventDefault();
          event.stopPropagation();
          this._isOpen = true;
          this.setState({
            open: true
          });
        }
      }
    }
  }, {
    key: '_addListeners',
    value: function _addListeners() {
      document.addEventListener('click', this._onDocumentClick);
    }
  }, {
    key: '_removeListeners',
    value: function _removeListeners() {
      document.removeEventListener('click', this._onDocumentClick);
    }
  }, {
    key: '_onDocumentClick',
    value: function _onDocumentClick(event) {
      if (this.state.open && !_reactDom2['default'].findDOMNode(this).contains(event.target)) {
        this._isOpen = false;
        this.setState({
          open: false
        });
      }
    }
  }, {
    key: '_isLeaf',
    value: function _isLeaf() {
      return !this.props.children || this.props.children.length === 0;
    }
  }, {
    key: '_onCloseChild',
    value: function _onCloseChild() {
      this._isOpen = false;
      this.setState({
        open: false
      });
      this.props.onClose();
    }
  }]);

  return MenuItem;
})(_reactEs6Component2['default']);

exports['default'] = MenuItem;
module.exports = exports['default'];
